Cody web recipes

philipp-spiess opened the pull request:
Refers to: #50678
Closes: #50787
Closes: #50788
Closes: #50791
Closes: #50678

This PR adds support for calling Cody recipes on the web by implementing an editor abstraction that can read the selection context from CodeMirror and with that it ships @almeidapaulooliveira's prototype for the recipe tooltip inside the blob view.

## Test plan

This one was too big for GitHub I reckon so here's a quick example uploaded to Google Drive: https://drive.google.com/file/d/1664-VZoLigTzqCsSypeTkiwLZ9MNdFYY/view?usp=sharing

<img width="1511" alt="Screenshot 2023-04-18 at 18 36 32" src="https://user-images.githubusercontent.com/458591/232845075-eff9b483-da7c-473c-866b-09b33a1329cd.png">


<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

philipp-spiess commented on client/web/src/cody/CodeMirrorEditor.ts:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    public getActiveTextEditorVisibleContent(): ActiveTextEditorVisibleContent | null {
+        return null
```
Still need to implement this

philipp-spiess commented on client/web/src/repo/blob/codemirror/tooltips/CodyTooltip.tsx:
```diff
@@ -0,0 +1,74 @@
+    })
+}
+
+// @TODO: Figure out how to properly update the component instead of recreating it every time
```
@taras-yemets / @fkling Can I pair with either of you on this? I’m a bit confused as to how exactly I can prevent CodeMirror to create a new tooltip on every selection change 😅

taras-yemets replied:
I would be happy to pair on it, @philipp-spiess!

sg-e2e-regression-test-bob commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.04% (+1.18 kb) | 0.15% (+22.22 kb) 🔺 | 0.18% (+21.04 kb) 🔺 | 0.40% (+3) 🔺 |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/ps/cody-web-recipes/compare-report.html;tab=live_object) for a full comparison between the commits 821cea3 and 2967c96 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

thenamankumar commented on client/cody-web/src/App.tsx:
```diff
@@ -12,6 +13,35 @@ import { useConfig } from './settings/useConfig'
+        return Promise.resolve()
+    },
+    async showQuickPick(labels) {
+        return window.prompt(`Choose: ${labels.join(', ')}`, labels[0]) || undefined
```
Add a TODO comment to fix UI for this later and put it in the github issue as well.

thenamankumar commented on client/cody-web/src/App.tsx:
```diff
@@ -12,6 +13,35 @@ import { useConfig } from './settings/useConfig'
+/* eslint-enable @typescript-eslint/require-await */
+
 export const App: React.FunctionComponent = () => {
     const [config, setConfig] = useConfig()
```
Why are we not using `useChatStore` here? Should we refactor it in this PR only?

philipp-spiess replied:
I didn't want to bring zustand and that pattern to the cody-web client yet (a standalone client). We may look into it as part of #50789

thenamankumar replied:
Where is this cody-web standalone client used right now?

philipp-spiess replied:
https://sourcegraph.slack.com/archives/C04NPH6SZMW/p1681908394374739

thenamankumar commented on client/web/src/cody/CodeMirrorEditor.ts:
```diff
@@ -0,0 +1,105 @@
+        return null
+    }
+
+    public getActiveTextEditorSelectionOrEntireFile(): ActiveTextEditorSelection | null {
```
None of the recipes right now don't need a fileName? Maybe we should also have `getActiveFile()`?

thenamankumar commented on client/web/src/repo/blob/codemirror/tooltips/CodyTooltip.tsx:
```diff
@@ -0,0 +1,74 @@
+ * Add a extension to CodeMirror extensions to display the Cody widget
+ * when some code is selected in the editor.
+ */
+export function codyWidgetExtension(): StateField<readonly Tooltip[]> {
```
Get CodyTooltip reviewed by someone having more experience with CodeMirror APIs.

philipp-spiess replied:
Yeah already tagged the pros here: https://github.com/sourcegraph/sourcegraph/pull/50728#discussion_r1170284961 😎

thenamankumar commented on client/web/src/stores/codyChat.ts:
```diff
@@ -1,46 +1,74 @@
+    initializeClient: (
+        config: Required<ClientInit['config']>,
+        editorStore: React.MutableRefObject<EditorStore>,
+        openCody: () => void
```
This store is part of `client/web`. It can have direct access to openCody without accepting it as a prop. 

Better to internalize this.

philipp-spiess replied:
the state for this currently lives in a component (`RepoContainer`) and is not global (zustand store). But yeah we can change that and make both globals and avoid some wiring.

philipp-spiess replied:
Oh I remember why I didn't do it right now. I use `useTemporarySetting()` for the cody state in the React tree (so that it is persisted across reloads).

I've moved the `useTemporarySetting()` hook inside the `useChatStore` for now though, to at least have it colocated.

philipp-spiess commented on client/web/src/repo/RepoContainer.tsx:
```diff
@@ -190,10 +190,12 @@ export const RepoContainer: FC<RepoContainerProps> = props => {
-
     const focusCodyShortcut = useKeyboardShortcut('focusCody')
     const [isCodySidebarOpen, setCodySidebarOpen] = useTemporarySetting('cody.showSidebar', false)
+    const openCody = useCallback(() => {
```
That makes sense! We can also do this in the `codyChat` store directly without it needing to go into the shared client

philipp-spiess commented:
@almeidapaulooliveira I've updated the selection UX like we discussed yesterday in the following ways which now feels much better to me. Want to check it out?

- The tooltip is now aligned to the left most point of the whole selection bounding rectangle. This was already the case for multiline selects but I've also changed single line selections now to align it to the start of the selection not the end. This is making it a bit more consistent between single and multi line
- The tooltip will now update _after the selection is complete_. This will reduce visual churn significantly (what we've discussed yesterday)

https://user-images.githubusercontent.com/458591/233059235-95d261e8-bbe4-49e3-9c4d-886c7189c1e0.mov

Big thanks to @taras-yemets for helping me get this fixed in CodeMirror.

taras-yemets commented on client/web/src/cody/CodeMirrorEditor.ts:
```diff
@@ -0,0 +1,107 @@
+            return null
+        }
+
+        const { from, to } = editor.view?.viewport
```
```suggestion
        const { from, to } = editor.view.viewport
```

taras-yemets commented on client/web/src/cody/widgets/components/RecipeAction.tsx:
```diff
@@ -0,0 +1,16 @@
+
+export interface RecipeActionProps {
+    title: string
+    onClick?: () => void
```
Do we expect actions without handlers?

taras-yemets commented on client/web/src/repo/blob/CodeMirrorBlob.tsx:
```diff
@@ -432,6 +435,21 @@ export const CodeMirrorBlob: React.FunctionComponent<BlobProps> = props => {
         }
     }, [])
 
+    useEffect(() => {
+        const view = editorRef.current
+        useEditorStore.setState({
+            editor: view
+                ? {
+                      view,
+                      repo: props.blobInfo.repoName,
+                      filename: props.blobInfo.filePath,
+                      content: props.blobInfo.content,
+                  }
+                : null,
+        })
+        return () => useEditorStore.setState({ editor: null })
+    }, [props.blobInfo.content, props.blobInfo.filePath, props.blobInfo.repoName])
+
```
A comment would be helpful for future us 🙂

philipp-spiess commented on client/web/src/cody/widgets/components/Recipes.tsx:
```diff
@@ -0,0 +1,18 @@
+import styles from './Recipes.module.scss'
+
+export const Recipes: React.FC<React.PropsWithChildren<{}>> = ({ children }) => {
+    const containerRef = useRef<HTMLDivElement>(null)
```
Leftover :)

taras-yemets commented on client/web/src/repo/blob/codemirror/tooltips/CodyTooltip.tsx:
```diff
@@ -0,0 +1,108 @@
+ * when some code is selected in the editor.
+ */
+export function codyWidgetExtension(): Extension {
+    return [codyTooltip, selectionChangedPlugin]
```
Do we need to initialize these extensions if Cody [is not enabled](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@3cadca99255123796d97da3c171e8e445cf983cd/-/blob/client/web/src/repo/RepoContainer.tsx?L192)?

philipp-spiess replied:
OH good point;

YES if cody sidebar is hidden, NO if cody feature flag is off. 😅

philipp-spiess replied:
Fixed

almeidapaulooliveira commented:
> @almeidapaulooliveira I've updated the selection UX like ...

Definitely better!! Thanks, @taras-yemets, for the help as well.

taras-yemets commented on client/web/src/repo/blob/codemirror/tooltips/CodyTooltip.tsx:
```diff
@@ -0,0 +1,108 @@
+    return [codyTooltip, selectionChangedPlugin]
+}
+
+const selectionChangedPlugin = ViewPlugin.fromClass(
+    class implements PluginValue {
+        constructor(public view: EditorView) {
+            document.body.addEventListener('mouseup', this.onPotentialSelectionChanged)
+            document.body.addEventListener('keyup', this.onPotentialSelectionChanged)
+        }
+        public destroy(): void {
+            document.body.removeEventListener('mouseup', this.onPotentialSelectionChanged)
+            document.body.removeEventListener('keyup', this.onPotentialSelectionChanged)
+        }
+        public onPotentialSelectionChanged = (): void => {
+            this.view.dispatch({ effects: [setCodyTooltip.of(computeCodyWidget(this.view.state))] })
+        }
+    }
+)
```
A comment about why we add event listeners to `document.body` instead of using CodeMirror's [`EditorView.domEventHandlers`](https://codemirror.net/docs/ref/#view.EditorView^domEventHandlers) would be helpful

almeidapaulooliveira commented on client/web/src/repo/blob/codemirror/tooltips/CodyTooltip.tsx:
```diff
@@ -1,42 +1,66 @@
 
-// @TODO: Figure out how to properly update the component instead of recreating it every time
-function getCodyWidget(state: EditorState): readonly Tooltip[] {
+const selectionChangedPlugin = ViewPlugin.fromClass(
```
Very interesting this plugin! 🧠

taras-yemets commented on client/web/src/repo/blob/codemirror/tooltips/CodyTooltip.tsx:
```diff
@@ -0,0 +1,108 @@
+    }
+}
+
+function isSelectionCollapsed(selection: EditorSelection): boolean {
+    if (!selection || selection.ranges.length === 0) {
+        return true
+    }
+
+    const [range] = selection.ranges
+    const { head, anchor } = range
+
+    // Don't show the tooltip if the selection is collapsed
+    if (head === anchor) {
+        return true
+    }
+
+    return false
+}
```
I wonder if `isSelectionCollapsed` return value is an equivalent of `selection.main.empty` value

taras-yemets commented on client/web/src/repo/blob/codemirror/tooltips/CodyTooltip.tsx:
```diff
@@ -0,0 +1,108 @@
+        return null
+    }
+
+    const [range] = selection.ranges
+    const { head, anchor } = range
+
+    const headLine = state.doc.lineAt(head)
+    const anchorLine = state.doc.lineAt(anchor)
+
+    let multiline = false
+    if (headLine.number !== anchorLine.number) {
+        multiline = true
+    }
+
+    // Find a position that is always the left most position of the selection bounding box
+    const pos = multiline
+        ? // When a multiline selection is made, the tooltip should be anchored to the start of the
+          // last line. This is to avoid the tooltip from jumping around too much
+          Math.max(headLine.from, anchorLine.from)
+        : // Otherwise, anchor the tooltip to the start of the selection
+          Math.min(head, anchor)
```
I wonder if it can be simplified to something like
```
    const lineFrom = state.doc.lineAt(selection.main.from)
    const lineTo = state.doc.lineAt(selection.main.to)
    const multiline = lineFrom.number !== lineTo.number
    const pos = multiline ? lineTo.from : selection.main.to
```

philipp-spiess replied:
Selections can be inverted, i.e they can start on the bottom and be moved upwards. This is why we have Math.max/Math.min here to make sure we still position stuff at the right start/end.

taras-yemets replied:
I guess that's why `from`/`to` and `anchor`/`head` fields exist on [SelectionRange](https://codemirror.net/docs/ref/#state.SelectionRange) (e.g. `from` is always the lower boundary).

philipp-spiess replied:
Oh, I see!

philipp-spiess replied:
Lol what I want can also be done in one line:

```ts
    const pos = state.doc.lineAt(selection.main.to).from
```

taras-yemets commented on client/web/src/cody/widgets/components/Recipe.tsx:
```diff
@@ -0,0 +1,49 @@
+}
+
+export function Recipe({ children, title, icon }: RecipeProps): JSX.Element {
+    const mainButtonRef = useRef<HTMLButtonElement>(null)
+    const [menuWidth, setMenuWidth] = useState<number | undefined>(undefined)
+
+    // Make the width ot the drop down menu the same of the main button.
+    useEffect(() => {
+        if (mainButtonRef.current) {
+            const width = mainButtonRef.current.offsetWidth
+            setMenuWidth(width)
+        }
+    }, [])
```
I assume we can't reach the same effect with CSS-only approach

philipp-spiess replied:
I haven't tried but I don't think that's the final UI anyways and we may tweak it in the next few days anyways.

taras-yemets reviewed (approved):
Great work, @philipp-spiess!
